CRM/Contribute/BAO/Contribution.php

```diff
@@ -5410,95 +5411,84 @@ protected static function isPaymentInstrumentChange(&$params, $pendingStatuses)
-ORDER BY  id DESC
-LIMIT 1;";
-        $dao = CRM_Core_DAO::executeQuery($sql);
-        if ($dao->fetch()) {
-          if (!empty($dao->membership_type_id)) {
-            $membershipParams['membership_type_id'] = $dao->membership_type_id;
-          }
-        }
-        $dao->free();
-
-        $membershipParams['num_terms'] = $contribution->getNumTermsByContributionAndMembershipType(
+        // Does the contact have a "current" membership (ie. not expired/pending etc).
+        $currentMembership = CRM_Member_BAO_Membership::getContactMembership($membershipParams['contact_id'],
           $membershipParams['membership_type_id'],
-          $primaryContributionID
+          $membershipParams['is_test'],
+          $membershipParams['id']
         );
-        // @todo remove all this stuff in favour of letting the api call further down handle in
-        // (it is a duplication of what the api does).
-        $dates = array_fill_keys(array(
-          'join_date',
-          'start_date',
-          'end_date',
-        ), NULL);
         if ($currentMembership) {
           /*
            * Fixed FOR CRM-4433
            * In BAO/Membership.php(renewMembership function), we skip the extend membership date and status
            * when Contribution mode is notify and membership is for renewal )
            */
           CRM_Member_BAO_Membership::fixMembershipStatusBeforeRenew($currentMembership, $changeDate);
-
-          // @todo - we should pass membership_type_id instead of null here but not
-          // adding as not sure of testing
-          $dates = CRM_Member_BAO_MembershipType::getRenewalDatesForMembershipType($membershipParams['id'],
-            $changeDate, NULL, $membershipParams['num_terms']
-          );
-          $dates['join_date'] = $currentMembership['join_date'];
         }
 
-        //get the status for membership.
-        $calcStatus = CRM_Member_BAO_MembershipStatus::getMembershipStatusByDate($dates['start_date'],
-          $dates['end_date'],
-          $dates['join_date'],
-          'today',
-          TRUE,
-          $membershipParams['membership_type_id'],
-          $membershipParams
-        );
-
-        unset($dates['end_date']);
-        $membershipParams['status_id'] = CRM_Utils_Array::value('id', $calcStatus, 'New');
-        //we might be renewing membership,
-        //so make status override false.
+        // Tell the Membership BAO to calculate membership status.
+        $membershipParams['skipStatusCal'] = 0;
+        $membershipParams['exclude_is_admin'] = TRUE;
         $membershipParams['is_override'] = FALSE;
         $membershipParams['status_override_end_date'] = 'null';
-
-        //CRM-17723 - reset static $relatedContactIds array()
-        // @todo move it to Civi Statics.
-        $var = TRUE;
-        CRM_Member_BAO_Membership::createRelatedMemberships($var, $var, TRUE);
-        civicrm_api3('Membership', 'create', $membershipParams);
       }
+
```

Comment:
@mattwire just wondering why have you moved this outside of the `if ($membership)` code

Reply:
Actually the `if ($membership)` is not required as the API will always return an array of values for a given membership ID and therefore `if($membership)` will always be true.  It made reviewing easier with it still in place (as a load of lines shift to the left by 2 spaces), but I have now removed the if statement entirely.